updated README (#963)
DAGWorks-Inc/hamilton: README.md

   <a href='https://hamilton.dagworks.io/en/latest/?badge=latest'>
      <img src='https://readthedocs.org/projects/hamilton/badge/?version=latest' alt='Documentation Status' />
   </a><a href="https://www.python.org/downloads/" target="_blank">
      <img src="https://img.shields.io/badge/python-3.8|%203.9|%203.10|%203.11|%203.12-blue.svg" alt="Python supported"/>
   </a>
   <a href="https://pypi.org/project/sf-hamilton/" target="_blank">
      <img src="https://badge.fury.io/py/sf-hamilton.svg" alt="PyPi Version"/>
   </a>
   <a href="https://pepy.tech/project/sf-hamilton" target="_blank">
      <img src="https://pepy.tech/badge/sf-hamilton" alt="Total Downloads"/>
   </a>

   <a target="_blank" href="https://linkedin.com/showcase/dagster" style="background:none">
      <img src="https://img.shields.io/badge/DAGWorks-Follow-purple.svg?logo=linkedin" />
   </a>
   <a href="https://join.slack.com/t/hamilton-opensource/shared_invite/zt-1bjs72asx-wcUTgH7q7QX1igiQ5bbdcg" target="_blank">
      <img src="https://img.shields.io/badge/Hamilton-Join-purple.svg?logo=slack" alt="Hamilton Slack"/>
   </a>

>    <a href="https://twitter.com/hamilton_os" target="_blank">
>       <img src="https://img.shields.io/badge/DAGWorks-Follow-purple.svg?logo=X"></src>
>    </a>

Reviewer: ```suggestion
   <a href="https://twitter.com/hamilton_os" target="_blank">
      <img src="https://img.shields.io/badge/HamiltonOS-Follow-purple.svg?logo=X"></src>
   </a>
   <a href="https://twitter.com/dagworks" target="_blank">
      <img src="https://img.shields.io/badge/DAGWorks-Follow-purple.svg?logo=X"></src>
   </a>
```

Reviewer: twitter link was to hamilton but it said dagworks. adding both.

</div>
<br></br>

> Hamilton is a general purpose Python library to author data transformations as a directed acyclic graph (DAG). It enables users to write ETL pipelines, ML workflows, LLM applications, RAG systems, business analytics dashboards, webscrapers, and more! The [Hamilton UI](https://hamilton.dagworks.io/en/latest/concepts/ui) allows to automatically visualize, catalog, and monitor execution.

Reviewer: ```suggestion
Hamilton is a general purpose Python library to author data transformations as a directed acyclic graph (DAG). It enables users to write ETL pipelines, ML workflows, LLM applications, RAG systems, business analytics dashboards, webscrapers, and more! The [Hamilton UI](https://hamilton.dagworks.io/en/latest/concepts/ui) allows you to automatically visualize, catalog, and monitor execution.
```

> Hamilton is a general purpose Python library to author data transformations as a directed acyclic graph (DAG). It enables users to write ETL pipelines, ML workflows, LLM applications, RAG systems, business analytics dashboards, webscrapers, and more! The [Hamilton UI](https://hamilton.dagworks.io/en/latest/concepts/ui) allows to automatically visualize, catalog, and monitor execution.

Reviewer 2: "general purpose python library to author" doesn't sell value. While we want it to be matter-of-fact, it should reflect/reiterate the tagline.

Hamilton uses regular Python functions to define DAGs; you don't even need to `import hamilton` ! It uses the function name and parameters to resolve the graph. As shown below, functions specify their dependencies creating readable code that can always be visualized.

<div align="center">
  <img src="./docs/_static/abc.png" alt="Create a project" width="70%"/>
</div>
<br>

> 🤖 Hamilton is great for DAGs. If you needs loops or conditional logic to create an LLM agent or a simulation take a look at our sister library  [Burr](https://github.com/dagworks-inc/burr).

# Installation

> Hamilton supports Python 3.8+. We include the optional `visualization` dependency to display our Hamilton DAG.

Reviewer: ```suggestion
Hamilton supports Python 3.8+. We include the optional `visualization` dependency to display our Hamilton DAG. Note: [graphviz](https://graphviz.org/download/) needs to be installed on your system independently for visualization.
```

> Hamilton supports Python 3.8+. We include the optional `visualization` dependency to display our Hamilton DAG.

Reviewer 2: Hmm, might want to note that we need graphviz?

```bash
pip install "sf-hamilton[visualization]"
```

To use the Hamilton UI, install the `ui` and `sdk` dependencies.

```bash

> pip install "sf-hamilton[ui,sdk]"

Reviewer: ```suggestion
pip install "sf-hamilton[ui,sdk]"
# then to start the UI
hamilton UI
```

Reviewer: oh i running it is further down.

```

To try Hamilton in the browser, visit [www.tryhamilton.dev](https://www.tryhamilton.dev/?utm_source=README)

> # Key Features

Reviewer 2: Capitalization is not consistent -- second word lowercase is what hte rrest of the post is

> # Key Features

Reviewer 2: "Why Hamilton?" might be a better name?

Data teams write code to deliver business value, but few have the resources to standardize practices and provide quality assurance. Moving from proof-of-concept to production, and cross-function collaboration (e.g., data science, engineering, ops) remain challenging for teams big or small.

> Hamilton bakes-in solutions to support you throughout a project's lifecycle:

Reviewer 2: "bakes-in" is awkard phrasing -- something simple like "Hamilton supports the entire lifecycle of data projects with:"

> - **Built-in coding style**. Hamilton is built around Python functions encouraging  modular, easy to read, self-documenting, and unit testable code.

Reviewer 2: Linking these to docs/resources might be nice

> - **Maintainable DAG definitions**.

Reviewer 2: This has no definition? I'd take away "built-in-coding-style" and use that definition for "Maintainable DAG definitions". Seems overly opinionated at first glance.

<p align="center">
  <img src="./docs/_static/hamilton_1.jpeg" alt="Description1" width="30%" style="margin-right: 20px;"/>
  <img src="./docs/_static/hamilton_2.jpeg" alt="Description2" width="30%" style="margin-right: 20px;"/>
  <img src="./docs/_static/hamilton_3.jpeg" alt="Description3" width="30%"/>
</p>
<p align="center">
  <em>DAG catalog, automatic dataset profiling, and execution tracking</em>
</p>

## Get started with the Hamilton UI

1. To use the Hamilton UI, install the dependencies (see `Installation` section) and start the server with

   ```bash
   hamilton ui
   ```

2. On the first connection, you will have to create a `username`. Then, create a new project and note the `project_id` (it should be `1`).

<div align="center">
  <img src="./docs/_static/new_project.png" alt="Create a project" width="70%"/>
</div>
<br>

> 3. Track your Hamilton DAG by create a `HamiltonTracker` object with your `username` and `project_id`. Now, all executions will be tracked and available from the UI!

Reviewer: ```suggestion
3. Track your Hamilton DAG by create a `HamiltonTracker` object with your `username` and `project_id`. Now, all DAGs, i.e. drivers, created and executed will be tracked and available from the UI!
```

   tracker = HamiltonTracker(
      username="my_username",
      project_id=1,
      dag_name="hello_world",
   )

   # building the `Driver` will add the DAG to the catalog
   dr = (
      driver.Builder()
      .with_modules(my_dag)
      .with_adapters(tracker)  # add your tracker here
      .build()
   )

   # executing the `Driver` will track results
   dr.execute(["C"])
   ```

# Documentation & learning resources

* 📚 See the [official documentation](https://hamilton.dagworks.io/) to learn about the core concepts of Hamilton.

* 👨‍🏫 Consult the [examples on GitHub](https://github.com/DAGWorks-Inc/hamilton/tree/main/examples) to learn about specific features or integrations with other frameworks.

* 📰 The [DAGWorks blog](https://blog.dagworks.io/) includes narrative tutorials and broader guides about how to build a data platform.

* 📺 Find video tutorials on the [DAGWorks YouTube channel](https://www.youtube.com/@DAGWorks-Inc)

* 📣 Reach out via the [Hamilton Slack community](https://join.slack.com/t/hamilton-opensource/shared_invite/zt-1bjs72asx-wcUTgH7q7QX1igiQ5bbdcg) for help and troubleshooting

# How does Hamilton compare to X?

> Hamilton is not an orchestrator ([you might not need one](https://blog.dagworks.io/p/lean-data-automation-a-principal)) nor a feature store ([but you can use it to build one!](https://blog.dagworks.io/p/featurization-integrating-hamilton)). It's closest to dbt; it helps you structure and manage data transformations. While dbt is focus on SQL and data tables, Hamilton supports the full Python ecosystem and is great for machine learning pipelines, working with documents & LLMs, web requests, and more.

Reviewer 2: This is good

> Hamilton is not an orchestrator ([you might not need one](https://blog.dagworks.io/p/lean-data-automation-a-principal)) nor a feature store ([but you can use it to build one!](https://blog.dagworks.io/p/featurization-integrating-hamilton)). It's closest to dbt; it helps you structure and manage data transformations. While dbt is focus on SQL and data tables, Hamilton supports the full Python ecosystem and is great for machine learning pipelines, working with documents & LLMs, web requests, and more.

Reviewer 2: OK, we don't want to lose people with assuming dbt knowledge. "If you know dbt -- hamilton does for python what DBT does for sql -- taking simliar (python-specific) standards and applying them to ..."

Another way to frame it is to think about the different layers of a data stack. Hamilton is at the **asset layer**. It helps you organize data transformations (the **expression layer**), manage changes, and validate & test data.

> <table>

Reviewer 2: Can you make this table full-width?

    <colgroup>
        <col style="width: 20%">
        <col style="width: 40%">
        <col style="width: 40%">
    </colgroup>
    <thead>
        <tr>
            <th>Layer</th>
            <th>Function</th>
            <th>Example Tools</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Orchestration</td>
            <td>Operational system for the creation of assets</td>
            <td>Airflow, Metaflow, Prefect, Kedro</td>
        </tr>
        <tr>
            <td>Asset</td>

>             <td>Organize expressions into meaningful units <br> (e.g., a dataset, a model)</td>

Reviewer 2: Links to tools?

>             <td><b>Hamilton</b>, dbt, dlt, SQLMesh, LangGraph</td>

Reviewer 2: We've gotta put in Burr if we put in langgraph

            <td>Perform data transformations</td>
            <td>Spark, Snowflake, DuckDB, RAPIDS</td>
        </tr>
        <tr>
            <td>Data</td>
            <td>Physical representation of data, inputs and outputs</td>
            <td>S3, Postgres, file system, Snowflake</td>
        </tr>
    </tbody>
</table>

>See our page on [Why use Hamilton?](https://hamilton.dagworks.io/en/latest/get-started/why-hamilton/) and framework [code comparisons](https://hamilton.dagworks.io/en/latest/code-comparisons/) for more information.

# 📑 License

Hamilton is released under the BSD 3-Clause Clear License. See [LICENSE](https://github.com/DAGWorks-Inc/hamilton/blob/main/LICENSE.md) for details.

# 🌎 Community
## 👨‍💻 Contributing
We take contributions, large and small. We operate via a [Code of Conduct](https://github.com/DAGWorks-Inc/hamilton/blob/main/CODE_OF_CONDUCT.md) and expect anyone
contributing to do the same.

To see how you can contribute, please read our [contributing guidelines](https://github.com/DAGWorks-Inc/hamilton/blob/main/CONTRIBUTING.md) and then our [developer
setup guide](https://github.com/DAGWorks-Inc/hamilton/blob/main/developer_setup.md).

## 😎 Used by

> Hamilton was started at Stitch Fix before the original creators founded DAGWorks Inc! It's a battle-tested framework that has been supporting production use cases since 2019.

Reviewer 2: Second `it's` is incorrect. First one is correct, but we should avoid them.

Read more about it's [origin story](https://multithreaded.stitchfix.com/blog/2021/10/14/functions-dags-hamilton/).

* [Stitch Fix](https://www.stitchfix.com/) — Time series forecasting
* [UK Government Digital Services](https://github.com/alphagov/govuk-feedback-analysis) — National feedback pipeline (processing & analysis)
* [IBM](https://www.ibm.com/) — Internal search and ML pipelines
* [Opendoor](https://www.opendoor.com/) — Manage PySpark pipelines
* [Lexis Nexis]() — Feature processing and lineage
* [Adobe]() — Prompt engineering research
* [British Cycling](https://www.britishcycling.org.uk/) — Telemetry analysis
* [Oak Ridge & PNNL](https://pnnl.gov/) — Naturf project
* [ORNL](https://www.ornl.gov/)
* [Federal Reserve Board](https://www.federalreserve.gov/)
* [Joby Aviation](https://www.jobyaviation.com/) — Flight data processing
* [Two](https://www.two.inc/)
* [Transfix](https://transfix.io/) — Online featurization and prediction
* [Railofy](https://www.railofy.com) — Orchestrate pandas code
* [Habitat Energy](https://www.habitat.energy/) — Time-series feature engineering
* [KI-Insurance](https://www.ki-insurance.com/) — Feature engineering

> * [Ascena Retail](https://www.ascena.com/) — Feature engineering

Reviewer 2: Wren AI?

* [NaroHQ](https://www.narohq.com/)
* [EquipmentShare](https://www.equipmentshare.com/)
* [Everstream.ai](https://www.everstream.ai/)
* [Flectere](https://flectere.net/)
* [F33.ai](https://f33.ai/)

## 🤝 Code Contributors
[![Contributors](https://contrib.rocks/image?repo=dagworks-inc/hamilton)](https://contrib.rocks/image?repo=dagworks-inc/hamilton)

## 🙌 Special Mentions & 🦟 Bug Hunters

> Thanks for our awesome community and their active involvement in the Hamilton library.

Reviewer 2: Replace with this:

```python
[Nils Olsson](https://github.com/nilsso), [Michał Siedlaczek](https://github.com/elshize), [Alaa Abedrabbo](https://github.com/AAbedrabbo), [Shreya Datar](https://github.com/datarshreya), [Baldo Faieta](https://github.com/baldofaieta), [Anwar Brini](https://github.com/AnwarBrini), [Gourav Kumar](https://github.com/gms101), [Amos Aikman](https://github.com/amosaikman), [Ankush Kundaliya](https://github.com/akundaliya), [David Weselowski](https://github.com/j7zAhU), [Peter Robinson](https://github.com/Peter4137), [Seth Stokes](https://github.com/sT0v), [Louis Maddox](https://github.com/lmmx), [Stephen Bias](https://github.com/s-ducks), [Anup Joseph](https://github.com/AnupJoseph), [Jan Hurst](https://github.com/janhurst), [Flavia Santos](https://github.com/flaviassantos), [Nicolas Huray](https://github.com/nhuray), [Manabu Niseki](https://github.com/ninoseki), [Kyle Pounder](https://github.com/kpounder), [Alex Bustos](https://github.com/bustosalex1), [Andy Day](https://github.com/adayNU)
```

- **Maintainable DAG definition**. Hamilton [automatically builds the DAG from a single line of code whether it has 10 or 1000 nodes](https://hamilton.dagworks.io/en/latest/concepts/driver/). It can also assemble multiple Python modules into a pipeline, encouraging modularity.

- **Expressive DAGs**. [Function modifiers](https://hamilton.dagworks.io/en/latest/concepts/function-modifiers/) are a unique feature to keep your code [DRY](https://en.wikipedia.org/wiki/Don't_repeat_yourself) and reduce the complexity of maintaining large DAGs. Other frameworks inevitably lead to code redundancy or bloated functions.

- **Portable transformations**. Your DAG is [independent of infrastructure or orchestration](https://blog.dagworks.io/publish/posts/detail/145543927?referrer=%2Fpublish%2Fposts), meaning you can develop and debug locally and reuse code across contexts (local, Airflow, FastAPI, etc.).

- **Separation of concerns**. Hamilton separates the DAG "definition" and "execution" which lets data scientists focus on solving problems and engineers manage production pipelines.

> - **Low-friction dev to prod**. Use `@config.when()` to modify your DAG between execution environments instead of error-prone `if/else` feature flags. The notebook extension prevents the pain of migrating code from a notebook to a Python module.

Author: I had it this way because it enabled me to link in order:
- functions and nodes
- Driver definition
- function modifiers

> - **Low-friction dev to prod**. Use `@config.when()` to modify your DAG between execution environments instead of error-prone `if/else` feature flags. The notebook extension prevents the pain of migrating code from a notebook to a Python module.

Reviewer 2: Don't think ordered linking is important -- we want to order it by clear attainable value IMO
